### PR TITLE
[libc] assert.h modifications for newhdrgen

### PR DESCRIPTION
**Pull request by aaryanshukla**
- currently assert.h has a lot of code that does not have a position in
  our yaml files. Other files like errno.h and syscall.h have this code
in their respective .h.def files. We want to add that feature to
assert.h by changing assert.h.def.

### CODE REVIEW

File: libc/include/assert.h.def [hunk 1]
```diff
@@ -12,4 +12,21 @@
 // This file may be usefully included multiple times to change assert()'s
 // definition based on NDEBUG.
 
+#ifndef __cplusplus
```
Comment (michaelrj-google) [inline libc/include/assert.h.def:15]: if you're adding this here it needs to be removed from the `api.td` files it's in (specifically baremetal, linux, and GPU).

**Review by michaelrj-google — COMMENTED** [review]
@jhuber6 is this okay for GPUs? Right now the new headergen design doesn't have support for different macros on different platforms.

File: libc/include/assert.h.def [hunk 2]
```diff
@@ -12,4 +12,21 @@
...
+#define static_assert _Static_assert
+#endif
+
+#undef assert
+#ifdef NDEBUG
+#define assert(e) (void)0
+#else
+#ifdef __cplusplus
+extern "C"
+#endif
```
Comment (petrhosek) [inline libc/include/assert.h.def:26]: This file already includes `__llvm-libc-common.h` so you can use `__BEGIN_C_DECLS` and `__END_C_DECLS`.

File: libc/include/assert.h.def [hunk 3]
```diff
@@ -12,4 +12,21 @@
 // This file may be usefully included multiple times to change assert()'s
 // definition based on NDEBUG.
 
+#ifndef __cplusplus
+#undef static_assert
+#define static_assert _Static_assert
```
Comment (petrhosek) [inline libc/include/assert.h.def:17]: This is already defined in `__llvm-libc-common.h` which is included above.

File: libc/newhdrgen/yaml/assert.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    standards:
+      - llvm_libc_ext
+    return_type: _Noreturn void
+    arguments:
+      - type: const char *
+      - type: const char *
+      - type: unsigned
+      - type: const char *
+    guard: __cplusplus
+
```
Comment (RoseZhang03) [inline libc/newhdrgen/yaml/assert.yaml:17]: Remove extra line here, but otherwise LGTM

Comment (aaryanshukla) [reply libc/newhdrgen/yaml/assert.yaml:17]: In the other Yaml files all of them have a newline at the end as well.

Comment (RoseZhang03) [reply libc/newhdrgen/yaml/assert.yaml:17]: If there are extra spaces on the newline, github will complain about it. If line is empty the line number should turn gray.

File: libc/include/assert.h.def [hunk 5]
```diff
@@ -12,4 +12,15 @@
 // This file may be usefully included multiple times to change assert()'s
 // definition based on NDEBUG.
 
+
+#undef assert
+#ifdef NDEBUG
+#define assert(e) (void)0
+#else
+
+_Noreturn void __assert_fail(const char *, const char *, unsigned, const char *) __NOEXCEPT;
```
Comment (michaelrj-google) [inline libc/include/assert.h.def:21]: this needs the `ifdef __cplusplus` guards so that it's properly `extern "C"` in C mode

**Review by jhuber6 — APPROVED** [review]

File: libc/config/linux/api.td [hunk 6]
```diff
@@ -9,42 +9,6 @@ include "spec/stdc_ext.td"
...
-    #define assert(e)  \
-      ((e) ? (void)0 : __assert_fail(#e, __FILE__, __LINE__, __PRETTY_FUNCTION__))
-
-    #endif
-  }];
-}
-
-def StaticAssertMacro : MacroDef<"static_assert"> {
-  let Defn = [{
-    #ifndef __cplusplus
```
Comment (michaelrj-google) [inline libc/config/linux/api.td:34]: this macro also needs to be added to the .h.def

**Pull request closed without merging**